src/tools/w3dviewer/emittercolorproppage.cpp

```diff
@@ -0,0 +1,480 @@
+            }
+
+            break;
+        }
+        case IDC_OPACITYSPIN: {
+            NMUPDOWN *nm = reinterpret_cast<NMUPDOWN *>(lParam);
+
+            if (hdr->code == UDN_DELTAPOS) {
+                m_opacityKeyFrames.Rand = (nm->iPos + nm->iDelta) / 255.0f;
+                m_instanceList->Set_Opacity_Keyframes(m_opacityKeyFrames);
+                SetModified();
+            }
+
+            break;
+        }
+    }
+
+    return CPropertyPage::OnNotify(wParam, lParam, pResult);
+}
+
+void EmitterColorPropPageClass::DoDataExchange(CDataExchange *pDX)
+{
+    DDX_Control(pDX, IDC_OPACITYSPIN, m_opacitySpin);
+    DDX_Control(pDX, IDC_RSPIN, m_redSpin);
+    DDX_Control(pDX, IDC_GSPIN, m_greenSpin);
+    DDX_Control(pDX, IDC_BSPIN, m_blueSpin);
+}
+
+BOOL EmitterColorPropPageClass::OnInitDialog()
+{
+    CDialog::OnInitDialog();
+    m_colorBar = static_cast<ColorBarClass *>(GetProp(::GetDlgItem(m_hWnd, IDC_COLORBAR), "CLASSPOINTER"));
+    m_opacityBar = static_cast<ColorBarClass *>(GetProp(::GetDlgItem(m_hWnd, IDC_OPACITYBAR), "CLASSPOINTER"));
+    m_colorBar->SetMinMax(0.0f, 1.0f);
+    m_opacityBar->SetMinMax(0.0f, 1.0f);
+    m_opacityBar->SetColor(0, 0.0f, 255.0f, 255.0f, 255.0f, 3);
+    m_opacitySpin.SetRange(0, 100);
+    m_redSpin.SetRange(0, 255);
+    m_greenSpin.SetRange(0, 255);
+    m_blueSpin.SetRange(0, 255);
+    m_opacitySpin.SetPos(m_oldOpacityKeyFrames.Rand * 100.0f);
+    m_redSpin.SetPos(m_oldColorKeyFrames.Rand.X * 255.0f);
+    m_greenSpin.SetPos(m_oldColorKeyFrames.Rand.Y * 255.0f);
+    m_blueSpin.SetPos(m_oldColorKeyFrames.Rand.Z * 255.0f);
+    m_colorBar->Clear();
+    m_opacityBar->Clear();
+    m_colorBar->SetColor(0,
+        0.0f,
+        m_oldColorKeyFrames.Start.X * 255.0f,
+        m_oldColorKeyFrames.Start.Y * 255.0f,
+        m_oldColorKeyFrames.Start.Z * 255.0f,
+        3);
+    m_opacityBar->SetColor(0,
+        0.0f,
+        m_oldOpacityKeyFrames.Start * 255.0f,
+        m_oldOpacityKeyFrames.Start * 255.0f,
+        m_oldOpacityKeyFrames.Start * 255.0f,
+        3);
+    m_colorBar->SetMinMax(0.0f, 1.0f);
+    m_opacityBar->SetMinMax(0.0f, 1.0f);
+
+    for (unsigned int i = 0; i < m_oldColorKeyFrames.NumKeyFrames; i++) {
+        m_colorBar->SetColor(i + 1,
+            m_oldColorKeyFrames.KeyTimes[i] / m_lifetime,
+            m_oldColorKeyFrames.Values[i].X * 255.0f,
+            m_oldColorKeyFrames.Values[i].Y * 255.0f,
+            m_oldColorKeyFrames.Values[i].Z * 255.0f,
+            3);
+    }
+
+    for (unsigned int i = 0; i < m_oldOpacityKeyFrames.NumKeyFrames; i++) {
+        m_colorBar->SetColor(i + 1,
+            m_oldOpacityKeyFrames.KeyTimes[i] / m_lifetime,
+            m_oldOpacityKeyFrames.Values[i] * 255.0f,
+            m_oldOpacityKeyFrames.Values[i] * 255.0f,
+            m_oldOpacityKeyFrames.Values[i] * 255.0f,
+            3);
+    }
+
+    m_colorKeyFrames.Rand = m_oldColorKeyFrames.Rand;
+    m_opacityKeyFrames.Rand = m_oldOpacityKeyFrames.Rand;

```

Comment:
This looks odd. Shouldn't it be the other way around after the update?

Reply:
No this is correct.